File: Exercise_1.py
# Problem1
# Search a 2D Matrix(https://leetcode.com/problems/search-a-2d-matrix/)
# // Time Complexity : O(logn)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach in three sentences only
class Solution:
    def search(self, nums, target):
        # checking if array is empty
        if len(nums) == 0:
            # not in nums
            return -1
        # taking two ptrs low
        low = 0
        # and high
        high = len(nums) - 1
        # while loop
        while (low <= high):
            # calculating mid
            mid = (low + high) // 2
            # if mid has target
            if nums[mid] == target:
                # target in nums return mid
                return mid
            # check if left sorted
            elif (nums[low] <= nums[mid]):
                # left sorted
                # check if target present in left part
                if (target >= nums[low] and target < nums[mid]):
                    # shorten our window by 1/2
                    high = mid - 1
                else:
                    low = mid + 1
            else:
                # right sorted
                if (target > nums[mid] and target <= nums[high]):
                    # shorten our window by 1/2
                    low = mid + 1
                else:
                    high = mid - 1
        # target not present
        return - 1


nums = [4, 5, 6, 7, 0, 1, 2]
target = 0
sol = Solution()
print(sol.search(nums=nums, target=target))

File: Exercise_1.cpp
// Time Complexity                                  : O(N)
// Space Complexity                                 : O(2*N) = O(N)
// Did this code successfully run on Leetcode       : Yes.
// Any problem you faced while coding this          : using char and string with maps in c++. I tried using something like - 
//                                                    unordered_map<char,string> mp but the comparison threw me some weird error. 
//                                                    But finally used this solution and it worked.

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    
    bool search(vector<vector<int>>& matrix,int row, int l, int r, int target) {
        
        while(l <= r) {
            int md = l + (r-l)/2;
            
            if(target == matrix[row][md]) return true;
            else if (target > matrix[row][md]){
                l = md + 1;
            } else {
                r = md - 1;
            }
        }
        
        return false;
    }
    
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        int n = matrix.size();
        int m = matrix[0].size() - 1;
        
        if(n == 1) {
            return search(matrix,0,0,m,target);
        }
        
        
        
        int l = 0;
        int r = n-1;
        
        while(l <= r) {
            int md = l + (r-l)/2;
            // cout << l << "," << r << "," << md << "\n";
            
            if(matrix[md][0] == target) {
                return true;
            } else if (target > matrix[md][0]) {
                
                if(target <= matrix[md][m]) {
                    return search(matrix,md,0,m,target);
                }
                
                l = md+1;
            } else {
                r = md - 1;
            }
        }
        
        return false;
        
    }
};